In [1]:
# Importing liabraries
import pandas as pd
import numpy as np

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
# Importing Datasets
books = pd.read_csv('Books.csv')
users = pd.read_csv('Users.csv')
rating = pd.read_csv('Ratings.csv')

In [4]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [5]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [6]:
rating.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

Checking the Shape of all datasets:

In [7]:
print(f"Shape of Books dataset are {books.shape}.")
print(f"Shape of Users dataset are {users.shape}.")
print(f"Shape of Ratings dataset are {rating.shape}.")

Shape of Books dataset are (271360, 8).
Shape of Users dataset are (278858, 3).
Shape of Ratings dataset are (1149780, 3).


Checking the Datatypes:

In [8]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271359 non-null  object
 3   Year-Of-Publication  271360 non-null  object
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: object(8)
memory usage: 16.6+ MB


In [9]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278858 non-null  int64  
 1   Location  278858 non-null  object 
 2   Age       168096 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


In [10]:
rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  1149780 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 26.3+ MB


Checking Null Values:

In [11]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [12]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [13]:
rating.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

Checking the Duplicate Values:

In [14]:
books.duplicated().sum()

0

In [15]:
users.duplicated().sum()

0

In [16]:
rating.duplicated().sum()

0

### 1. Pupularity Based Recommender System:

In [17]:
books_ratings = rating.merge(books,on = 'ISBN')
books_ratings.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X            9  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   
2                2002  Ballantine Books   
3                2002  Ballantine Books   
4                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

Number of Ratings on Each Book:

In [18]:
num_of_ratings = books_ratings.groupby('Book-Title')['Book-Rating'].count().reset_index()
num_of_ratings.rename(columns={'Book-Rating':'Number of Ratings'},inplace=True)
num_of_ratings.head()

Book-Title  Number of Ratings
0   A Light in the Storm: The Civil War Diary of ...                  4
1                              Always Have Popsicles                  1
2               Apple Magic (The Collector's series)                  1
3   Ask Lily (Young Women of Faith: Lily Series, ...                  1
4   Beyond IBM: Leadership Marketing and Finance ...                  1

Average Ratings on Each Book:

In [19]:
avg_of_ratings = books_ratings.groupby('Book-Title')['Book-Rating'].mean().reset_index()
avg_of_ratings.rename(columns={'Book-Rating':'Average of Ratings'},inplace=True)
avg_of_ratings.head()

Book-Title  Average of Ratings
0   A Light in the Storm: The Civil War Diary of ...                2.25
1                              Always Have Popsicles                0.00
2               Apple Magic (The Collector's series)                0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...                8.00
4   Beyond IBM: Leadership Marketing and Finance ...                0.00

Merge Number of Ratings and Average Ratings DF

In [20]:
popular_df = num_of_ratings.merge(avg_of_ratings, on = 'Book-Title')
popular_df.head()

Book-Title  Number of Ratings  \
0   A Light in the Storm: The Civil War Diary of ...                  4   
1                              Always Have Popsicles                  1   
2               Apple Magic (The Collector's series)                  1   
3   Ask Lily (Young Women of Faith: Lily Series, ...                  1   
4   Beyond IBM: Leadership Marketing and Finance ...                  1   

   Average of Ratings  
0                2.25  
1                0.00  
2                0.00  
3                8.00  
4                0.00

So for Popularity Based Recommender System We will display the top 50 books on with atleast 250 ratings:

In [21]:
popular_df = popular_df[popular_df['Number of Ratings'] >= 250].sort_values(by='Average of Ratings',ascending=False)

In [22]:
popular_df.head()

Book-Title  Number of Ratings  \
80434  Harry Potter and the Prisoner of Azkaban (Book 3)                428   
80422       Harry Potter and the Goblet of Fire (Book 4)                387   
80441     Harry Potter and the Sorcerer's Stone (Book 1)                278   
80426  Harry Potter and the Order of the Phoenix (Boo...                347   
80414   Harry Potter and the Chamber of Secrets (Book 2)                556   

       Average of Ratings  
80434            5.852804  
80422            5.824289  
80441            5.737410  
80426            5.501441  
80414            5.183453

Now we need the Authour Name, Publisher Name and Picture link So we will merge it with BOOKS dataset.

In [23]:
popular_df.merge(books,on="Book-Title").drop_duplicates('Book-Title')

Book-Title  Number of Ratings  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)                428   
3         Harry Potter and the Goblet of Fire (Book 4)                387   
5       Harry Potter and the Sorcerer's Stone (Book 1)                278   
9    Harry Potter and the Order of the Phoenix (Boo...                347   
13    Harry Potter and the Chamber of Secrets (Book 2)                556   
..                                                 ...                ...   
716       Vinegar Hill (Oprah's Book Club (Paperback))                265   
717                                           Whispers                286   
727                                  Presumed Innocent                294   
733                                       Isle of Dogs                288   
739                                        Wild Animus               2502   

     Average of Ratings        ISBN        Book-Author Year-Of-Publication  \
0              5.852804  0439136350      J. K. Rowling                1999   
3              5.824289  0439139597      J. K. Rowling                2000   
5              5.737410  0590353403      J. K. Rowling                1998   
9              5.501441  043935806X      J. K. Rowling                2003   
13             5.183453  0439064872      J. K. Rowling                2000   
..                  ...         ...                ...                 ...   
716            2.245283  0380730138   A. Manette Ansay                1998   
717            2.199301  0440216745        BELVA PLAIN                1994   
727            2.139456  0446359866        Scott Turow                1989   
733            2.000000  0425182908  Patricia Cornwell                2002   
739            1.019584  0971880107       Rich Shapero                2004   

                    Publisher  \
0                  Scholastic   
3                  Scholastic   
5                  Scholastic   
9                  Scholastic   
13                 Scholastic   
..                        ...   
716                 Perennial   
717                      Dell   
727              Warner Books   
733  Berkley Publishing Group   
739                   Too Far   

                                           Image-URL-S  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
716  http://images.amazon.com/images/P/0380730138.0...   
717  http://images.amazon.com/images/P/0440216745.0...   
727  http://images.amazon.com/images/P/0446359866.0...   
733  http://images.amazon.com/images/P/0425182908.0...   
739  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
716  http://images.amazon.com/images/P/0380730138.0...   
717  http://images.amazon.com/images/P/0440216745.0...   
727  http://images.amazon.com/images/P/0446359866.0...   
733  http://images.amazon.com/images/P/0425182908.0...   
739  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
5    http://images.amazon.com/images/P/0590353403.0...  
9    http://images.amazon.com/images/P/043935806X.0...  
13   http://images.amazon.com/images/P/0439064872.0...  
..                                               

In [24]:
final_popular_df = popular_df.merge(books,on="Book-Title").drop_duplicates('Book-Title')[['Book-Title','Book-Author','Image-URL-M','Number of Ratings','Average of Ratings']]

In [25]:
final_popular_df

Book-Title        Book-Author  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)      J. K. Rowling   
3         Harry Potter and the Goblet of Fire (Book 4)      J. K. Rowling   
5       Harry Potter and the Sorcerer's Stone (Book 1)      J. K. Rowling   
9    Harry Potter and the Order of the Phoenix (Boo...      J. K. Rowling   
13    Harry Potter and the Chamber of Secrets (Book 2)      J. K. Rowling   
..                                                 ...                ...   
716       Vinegar Hill (Oprah's Book Club (Paperback))   A. Manette Ansay   
717                                           Whispers        BELVA PLAIN   
727                                  Presumed Innocent        Scott Turow   
733                                       Isle of Dogs  Patricia Cornwell   
739                                        Wild Animus       Rich Shapero   

                                           Image-URL-M  Number of Ratings  \
0    http://images.amazon.com/images/P/0439136350.0...                428   
3    http://images.amazon.com/images/P/0439139597.0...                387   
5    http://images.amazon.com/images/P/0590353403.0...                278   
9    http://images.amazon.com/images/P/043935806X.0...                347   
13   http://images.amazon.com/images/P/0439064872.0...                556   
..                                                 ...                ...   
716  http://images.amazon.com/images/P/0380730138.0...                265   
717  http://images.amazon.com/images/P/0440216745.0...                286   
727  http://images.amazon.com/images/P/0446359866.0...                294   
733  http://images.amazon.com/images/P/0425182908.0...                288   
739  http://images.amazon.com/images/P/0971880107.0...               2502   

     Average of Ratings  
0              5.852804  
3              5.824289  
5              5.737410  
9              5.501441  
13             5.183453  
..                  ...  
716            2.245283  
717            2.199301  
727            2.139456  
733            2.000000  
739            1.019584  

[186 rows x 5 columns]

So these are our most popular books with atleast 250+ Number of Ratings

In [26]:
import pickle
pickle.dump(final_popular_df,open('popular.pkl','wb'))

### 2. Collaborative Filtering Based Recommender System

So in this type of recommendation system we will recommend the books on the base of the ratings from the users:
1. We will only consider those users that rated atleast 200 books and those books that got atleast ratings from 50 users.
2. Users will be our features and Books-Name will be our index

In [27]:
books_ratings

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1           2313  034545104X            5   
2           6543  034545104X            0   
3           8680  034545104X            5   
4          10314  034545104X            9   
...          ...         ...          ...   
1031131   276688  0517145553            0   
1031132   276688  1575660792            7   
1031133   276690  0590907301            0   
1031134   276704  0679752714            0   
1031135   276704  0806917695            5   

                                                Book-Title      Book-Author  \
0                                     Flesh Tones: A Novel       M. J. Rose   
1                                     Flesh Tones: A Novel       M. J. Rose   
2                                     Flesh Tones: A Novel       M. J. Rose   
3                                     Flesh Tones: A Novel       M. J. Rose   
4                                     Flesh Tones: A Novel       M. J. Rose   
...                                                    ...              ...   
1031131                                    Mostly Harmless    Douglas Adams   
1031132                                        Gray Matter  Shirley Kennett   
1031133  Triplet Trouble and the Class Trip (Triplet Tr...     Debbie Dadey   
1031134  A Desert of Pure Feeling (Vintage Contemporaries)   Judith Freeman   
1031135  Perplexing Lateral Thinking Puzzles: Scholasti...      Paul Sloane   

        Year-Of-Publication                          Publisher  \
0                      2002                   Ballantine Books   
1                      2002                   Ballantine Books   
2                      2002                   Ballantine Books   
3                      2002                   Ballantine Books   
4                      2002                   Ballantine Books   
...                     ...                                ...   
1031131                1995             Random House Value Pub   
1031132                1996  Kensington Publishing Corporation   
1031133                1997                              Apple   
1031134                1997                  Vintage Books USA   
1031135                1997                Sterling Publishing   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-L  
0        http://images.amazon.com/images/P/034545104X.0...  
1        http://images.amazon.com/images/P/034545104X.0...  
2        http://images.amazon.com/images/P/034545104X

Getting the users and how many times they rate the book(On Base of Rating):

In [28]:
no_of_rating_users = books_ratings.groupby('User-ID')['Book-Rating'].count()>200
no_of_rating_users = no_of_rating_users[no_of_rating_users].index

In [29]:
filtered_ratings = books_ratings[books_ratings['User-ID'].isin(no_of_rating_users)]
filtered_ratings.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
2      6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
5     23768  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
7     28523  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
15    77940  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
16    81977  034545104X            0  Flesh Tones: A Novel  M. J. Rose   

   Year-Of-Publication         Publisher  \
2                 2002  Ballantine Books   
5                 2002  Ballantine Books   
7                 2002  Ballantine Books   
15                2002  Ballantine Books   
16                2002  Ballantine Books   

                                          Image-URL-S  \
2   http://images.amazon.com/images/P/034545104X.0...   
5   http://images.amazon.com/images/P/034545104X.0...   
7   http://images.amazon.com/images/P/034545104X.0...   
15  http://images.amazon.com/images/P/034545104X.0...   
16  http://images.amazon.com/images/P/034545104X.0...   

                                          Image-URL-M  \
2   http://images.amazon.com/images/P/034545104X.0...   
5   http://images.amazon.com/images/P/034545104X.0...   
7   http://images.amazon.com/images/P/034545104X.0...   
15  http://images.amazon.com/images/P/034545104X.0...   
16  http://images.amazon.com/images/P/034545104X.0...   

                                          Image-URL-L  
2   http://images.amazon.com/images/P/034545104X.0...  
5   http://images.amazon.com/images/P/034545104X.0...  
7   http://images.amazon.com/images/P/034545104X.0...  
15  http://images.amazon.com/images/P/034545104X.0...  
16  http://images.amazon.com/images/P/034545104X.0...

Filtering on Base of Books with >50 ratings

In [30]:
ratings_on_books = filtered_ratings.groupby('Book-Title')['Book-Rating'].count() >=50
ratings_on_books = ratings_on_books[ratings_on_books].index

In [31]:
ratings_on_books

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)

In [32]:
filtered_ratings[filtered_ratings['Book-Title'].isin(ratings_on_books)]

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

In [33]:
final_ratings = filtered_ratings[filtered_ratings['Book-Title'].isin(ratings_on_books)]
final_ratings

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

In [34]:
pt = final_ratings.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [35]:
pt.shape

(706, 810)

In [36]:
pt.fillna(0,inplace=True)

In [37]:
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

Now we can represent each book with 810 users and each of then form a vector.<br>
Like book1 = [user1,user2,......,user810] so we will calculate the Eucledian distance b/w each vetcor and on base of that distance we will recommend the book we nearest distance.

In [38]:
from sklearn.metrics.pairwise import cosine_similarity

In [39]:
similarity_score = cosine_similarity(pt)
similarity_score

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

Now build a function and get suggestions of recommended books

In [40]:
print(np.where(pt.index=='1984'))#feteching index
print(np.where(pt.index=='1984')[0][0])

(array([0], dtype=int64),)
0


In [41]:
#Getting the top 5 books with max similarity(using enumerate to make index)
sorted(list(enumerate(similarity_score[0])),key=lambda x: x[1],reverse=True)[1:6]

[(47, 0.2702651417103732),
 (545, 0.26396193711234966),
 (82, 0.2366937434740099),
 (634, 0.23299389358170397),
 (551, 0.2262639743141286)]

In [52]:
def recommend(book_name):
    #fetching index
    index = np.where(pt.index==book_name)[0][0]
    similar_items = sorted(list(enumerate(similarity_score[index])),key=lambda x: x[1],reverse=True)[1:4]
    data = []
    for i in similar_items:
        item = []
        #print(i[0]) # It will return the index of similar books
        #print(pt.index[i[0]]) # it will return the names of similar books
        temp_df = books[books['Book-Title'] == pt.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Title'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Author'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Image-URL-M'].values))
        
        data.append(item)
    return data

In [57]:
recommend('The Hobbit : The Enchanting Prelude to The Lord of the Rings')[0][2]

'http://images.amazon.com/images/P/0345339711.01.MZZZZZZZ.jpg'

In [54]:
pickle.dump(pt,open('pt.pkl','wb'))
pickle.dump(books,open('books.pkl','wb'))
pickle.dump(similarity_score,open('similarity_score.pkl','wb'))